book/python/using-connectors.md

```diff
@@ -0,0 +1,191 @@
+
+### Redis
+
+Redis has many ways to be used as a source and a sink. We've provided two starting points to show how a source and sink can work. The script is very easy to modify and uses the well maintained [redis library](https://pypi.org/project/redis/).
+
+For the source, we show an example using the subscription support offered by Redis pubsub and it's a natural fit for many stream processing applications. Check out `redis_subscriber_source` for [details](https://github.com/WallarooLabs/wallaroo/tree/{{book.wallaroo_version}}/connectors/redis_subscriber_source).
+
+For the sink, we show that not everything needs to look exactly like a stream to be used with Wallaroo. We use a give hash in redis and set key, value pairs on that hash for each tuple passed to the sink. This is very handy for those cases that want to keep track of the latest value. See `redis_hash_sink` for [details](https://github.com/WallarooLabs/wallaroo/tree/{{book.wallaroo_version}}/connectors/redis_hash_sink).
+
+### RabbitMQ
+
+Our RabbitMQ source uses the [pika library](https://pypi.org/project/pika/). We support autmatic provisioning of missing queues in our setup code but it's advised that users ensure they provision their broker's exchange with the appropriate type before hand.
+
+### UDP
+
+Our UDP support uses the Python standard library. To see an example of it in action, check out the [celsius_connectors example](https://github.com/Wallaroo/wallaroo/tree/{{book.wallaroo_version}}/examples/python/celsius_connectors/).
+
+### S3
+
+This sink is powered by the [boto3 library](https://pypi.org/project/boto3/) and allows key value pairs to be uploaded as named blobs to an S3 bucket.
+
+### Postgres
+
+We do offer a some starting templates for postgres integration. These depend on your database schema and so we've left them in the [connectors/templates directory](https://github.com/WallarooLabs/wallaroo/tree/{{book.wallaroo_version}}/connectors/templates).
+
+As a source we have a very interesting example of how notifiers can turn a regular table into a stream. If you're struggling with finding a efficient way to stream relational data, this might be a good place to start.
+
+On the sink side, we support inserting new records for each sink output. This should be easy to modify to allow more complex database interactions.
+
+## Building Your Own Connectors
+
+While we offer prebuild connectors and they can be modified, somethins you might find yourself needing to build a connector from scratch. We'll walk through a source and sink example.
+
+### Custom Source Connector
+
+To build a new source, you need to import `wallaroo.experimental` along with any other runtime requirements you have. Because this is a vanilla python runtime, you're welcome to organize your code however you like.
+
+When your script starts running, you need to create the proper
+connector instance type. In this case we'd like to build a source connctor so we'll use this:
```

Comment:
"source connector"

Reply:
You're convincing me that I need to install a spell check module for my code editor now. Thanks!